src/Components/src/Microsoft.AspNetCore.Components.Browser/Hosting/WebAssemblyHostBuilder.cs

```diff
@@ -85,7 +85,23 @@ private void CreateServiceProvider()
                 configureServicesAction(_BrowserHostBuilderContext, services);
             }
 
-            _appServices = services.BuildServiceProvider();
+            _appServices = GetProviderFromFactory(services);
+
+            IServiceProvider GetProviderFromFactory(IServiceCollection collection)
+            {
+                var provider = collection.BuildServiceProvider();
+                var factory = provider.GetService<IServiceProviderFactory<IServiceCollection>>();
+
+                if (factory != null)
+                {
+                    using (provider)
+                    {
+                        return factory.CreateServiceProvider(factory.CreateBuilder(collection));
+                    }
+                }
+
+                return provider;
+            }
```

Comment:
@davidfowl @tratcher is this right? I thought we exposed the service provider factory through an extra method on the hosting builder.

Think of this code as aligning with the requirements of generic host in 3.0 rather than web host.

Reply:
You don't want to build the container twice. See:
https://github.com/aspnet/Extensions/blob/ea0df662ab06848560d85545f3443964c57e9318/src/Hosting/Hosting/src/HostBuilder.cs#L198-L205
And 
https://github.com/aspnet/Extensions/blob/ea0df662ab06848560d85545f3443964c57e9318/src/Hosting/Hosting/src/HostBuilder.cs#L79-L89

Reply:
OK, thanks for the info!

I've updated it to follow the same patterns as in those examples. It seems to fit in fine, since the web assembly host builder has direct equivalents to each concept. Does the updated code look right to you, @Tratcher and @rynowak?

If there are any deeper complications, we don't strictly have to do any of this PR at all. It's not been a widely-requested feature, and could be re-attempted at some later date if we want.